lib/Sema/TypeCheckExprObjC.cpp

```diff
@@ -317,6 +317,8 @@ Optional<Type> TypeChecker::checkObjCKeyPathExpr(DeclContext *dc,
     // Handle property references.
     if (auto var = dyn_cast<VarDecl>(found)) {
       validateDecl(var);
+      if (var->isInvalid() || !var->hasInterfaceType())
```

Comment:
Try a property with no declared type and an initial value expression that references itself via a keypath

Reply:
To be clear, I can get this code to be hit, but not with a nullptr interface type. Instead, we have an error type.

I'll push another commit that checks explicitly for error types and adds a test case that exercises that path.

It would be nice to understand how this was actually reaching this point with a null type, though - it's probably indicative of some bad path in decl validation.